### PR TITLE
Transverse filed (TF) fit functions

### LINKED ISSUE #22423: TF asymmetry function wrapper
Either converts a function, f, into one for calculating the TF normalisation 

N(1+f)

or extracts f from the above equation

### PR DESCRIPTION
**Pull request by AnthonyLim23**
**Description of work**
The Transverse field (TF) fit function is used to calculate the normalisation of muon asymmetry data. For this PR I have created an algorithm that takes the user function f (e.g. a sinusoidal function) and wrap it for the normalisation:

N(1+f)+expMuonDecay

By default we set the amplitude of ExpMuonDecay to zero. N is the normalisation. 

The extract mode will take the above function and return the user function f. 
 
**To test:**
There are a few scripts below that depict the use of the algorithm. 

1D example, the fixes and ties should at the end and in `fittedFunction `
```
#create a normalisation table
tab = CreateEmptyTableWorkspace()
tab.addColumn('double', 'norm')
tab.addColumn('str', 'name')
tab.addColumn('str', 'method')

tab.addRow([1.,"a","Estimate"])
tab.addRow([2.,"Run;;Group;;b;;Asym;;#1","Estimate"])
tab.addRow([3.,"Run;;Group;;c;;Asym;;#1","Estimate"])
tab.addRow([4.,"Run;;Group;;d;;Asym;;#1","Estimate"])

#create original function and workspace
myFunc='name=LinearBackground,A0=3,A1=4;name=LinearBackground,A0=0,A1=2;ties=(f0.A1=3, f0.A0=f1.A0)'
ws = CreateWorkspace(DataX=[1,2,3,4,5,6,7], DataY=[1,2,3,4,5,6],OutputWorkspace="a")

TFFunc = ConvertFitFunctionForMuonTFAsymmetry(myFunc,tab,["a"],"Construct")

# do a fit with new function

fit =mantid.AlgorithmManager.create("Fit")
fit.setProperty("Function",str(TFFunc))
fit.setProperty("InputWorkspace",ws)
fit.setProperty("Output",'fitWS')
fit.execute()
fittedFunc = fit.getPropertyValue("Function")

returnFunc = ConvertFitFunctionForMuonTFAsymmetry(str(fittedFunc),tab,["a"],"Extract")
# 0 iteration fit to get param table -> wont change function values
fit_output = Fit(Function=str(returnFunc),InputWorkspace=ws,MaxIterations=0,Output="return_params")

paramTable = fit_output.OutputParameters  # table containing the optimal fit parameters
print('Gradients (shared):')
print("Gradient 1: {0:.2f}".format(paramTable.column(1)[3]))
print("Gradient 2: {0:.2f}".format(paramTable.column(1)[1]))
print('offsets:')
print("Constant 1: {0:.2f}".format(paramTable.column(1)[0]))
print("Constant 2: {0:.2f}".format(paramTable.column(1)[2]))
```

Multi fitting example:

```
import math
import numpy as np

# create data
xData=np.linspace(start=0,stop=10,num=22)
yData=[]
for x in xData:
    yData.append(2.0*x+10.)
yData2=[]
for x in xData:
    yData2.append(5.0*x+7.)
# create workspaces
input = CreateWorkspace(xData,yData,OutputWorkspace="Run; Group; a; Asym; #1")
input2 = CreateWorkspace(xData,yData2,OutputWorkspace="Run; Group; b; Asym; #1")
# create function
myFunc=';name=LinearBackground,$domains=i,A0=0,A1=1'
myFunc2=';name=LinearBackground,$domains=i,A0=2,A1=3'
multiFunc='composite=MultiDomainFunction,NumDeriv=1'+myFunc+myFunc2 +';ties=(f0.A1=f1.A1,f0.A0=1.23)'


tab = CreateEmptyTableWorkspace()
tab.addColumn('double', 'norm')
tab.addColumn('str', 'name')
tab.addColumn('str', 'method')

tab.addRow([1.,"Run;;Group;;a;;Asym;;#1","Estimate"])
tab.addRow([2.,"Run;;Group;;b;;Asym;;#1","Estimate"])
tab.addRow([3.,"Run;;Group;;c;;Asym;;#1","Estimate"])
tab.addRow([4.,"Run;;Group;;d;;Asym;;#1","Estimate"])



bob= ConvertFitFunctionForMuonTFAsymmetry(multiFunc,tab,["Run; Group; a; Asym; #1","Run; Group; b; Asym; #1","Run; Group; b; Asym; #1"],"Construct")

print "boib"
print bob
print
out = Fit(Function=multiFunc, InputWorkspace=input, WorkspaceIndex=0, \
                 InputWorkspace_1=input2, WorkspaceIndex_1=0, \
                 StartX = 0.1, EndX=9.5, StartX_1 = 0.1, EndX_1=9.5,Output='fit_original')


fit =mantid.AlgorithmManager.create("Fit")
fit.setProperty("Function",str(bob))
fit.setProperty("InputWorkspace",input)
fit.setProperty("WorkspaceIndex",0)
fit.setProperty("InputWorkspace_1",input2)
fit.setProperty("WorkspaceIndex_1",0)
fit.setProperty("StartX" ,0.1)
fit.setProperty("EndX",9.5)
fit.setProperty("StartX_1" ,0.1)
fit.setProperty("EndX_1",9.5)
fit.setProperty("Output",'fit2')
fit.execute()
a = fit.getPropertyValue("Function")

print a
b=ConvertFitFunctionForMuonTFAsymmetry(a,tab,["Run; Group; a; Asym; #1","Run; Group; b; Asym; #1"],"Extract")
print
print b
fit_output = Fit(Function=b, InputWorkspace=input, WorkspaceIndex=0, \
                 InputWorkspace_1=input2, WorkspaceIndex_1=0, \
                 StartX = 0.1, EndX=9.5, StartX_1 = 0.1, EndX_1=9.5,Output='fit')
paramTable = fit_output.OutputParameters  # table containing the optimal fit parameters
# print results
print('Gradients (shared):')
print("Gradient 1: {0:.2f}".format(paramTable.column(1)[3]))
print("Gradient 2: {0:.2f}".format(paramTable.column(1)[1]))
print('offsets:')
print("Constant 1: {0:.2f}".format(paramTable.column(1)[0]))
print("Constant 2: {0:.2f}".format(paramTable.column(1)[2]))
```

Fixes #22423. 

**Release Notes** 

In release notes
---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by martyngigg** [comment]
@AnthonyLim23 It looks like there is now a conflict in `QENSFitSequential`

**Review by martyngigg — COMMENTED** [review]
`Framework/Muon/inc/MantidMuon/CMakeLists.txt` has snuck in: can we delete this?

File: Framework/Muon/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+set ( INC_FILES
+	inc/MantidMuon/ConvertFitFunctionForMuonTFAsymmetry.h
+
+	#inc/MantidMuon/ApplyMuonDetectorsGrouping.h
+	#inc/MantidMuon/MuonAlgorithmHelper.h
+)
+
+# set(SRC_UNITY_IGNORE_FILES)
```
Comment (martyngigg) [inline Framework/Muon/CMakeLists.txt]: Can we remove the commented out code?

File: Framework/Muon/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+#if(UNITY_BUILD)
+#  include(UnityBuild)
+#  enable_unity_build(Muon SRC_FILES C_SRC_FILES SRC_UNITY_IGNORE_FILES 10)
+#endif(UNITY_BUILD)
+
+ set ( TEST_FILES 
+        ConvertFitFunctionForMuonTFAsymmetryTest.h)
+
+# set ( TEST_PY_FILES )
```
Comment (martyngigg) [inline Framework/Muon/CMakeLists.txt]: Same here.

File: Framework/Muon/inc/MantidMuon/ConvertFitFunctionForMuonTFAsymmetry.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+private:
+  // Overridden Algorithm methods
+  void init() override;
+  void exec() override;
+  std::map<std::string, std::string> validateInputs() override;
+  Mantid::API::IFunction_sptr
+  getTFAsymmFitFunction(const Mantid::API::IFunction_sptr &original,
+                        const std::vector<double> norms);
```
Comment (martyngigg) [inline Framework/Muon/inc/MantidMuon/ConvertFitFunctionForMuonTFAsymmetry.h]: Can we take this by reference `const std::vector<double> &`?

**Comment by AnthonyLim23** [comment]
trying to fix the merge conflict, having some difficulties picking up the file. I will delete the extra cmake

**Review by martyngigg — COMMENTED** [review]
` Framework/Muon/src/CMakeLists.txt` is also empty and not required - can we remove it?

File: Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+  auto LHName = stringTie.substr(0, seperator);
+  LHName = trimTie(LHName);
+  // this one includes = sign
+  auto RHName = stringTie.substr(seperator, std::string::npos);
+  RHName = trimTie(RHName);
+
+  return LHName + RHName;
+}
+
+template <typename T1>
```
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp]: your right that can be just a string

File: Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp [hunk 5]
```diff
@@ -46,8 +46,7 @@ std::string rmInsertFunction(const std::string &originalTie) {
   return LHName + RHName;
 }
 
-template <typename T1>
-int findName(const std::vector<std::string> &colNames, const T1 &name) {
+int findName(const std::vector<std::string> &colNames, std::string &name) {
```
Comment (martyngigg) [inline Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp]: In this case you could just go fo `findName(const std::vector<std::string> &colNames, std::string name)` and take the string by value. When you call it you can do
```
auto wsNamesIndex = findName(colNames, "name");
```
and the `std::string` will move constructed as a `std::string` is able to be constructed from a `const char *`.

**Review by martyngigg — APPROVED** [review]
I'm happy with this as soon as the builds pass.

File: Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp [hunk 6]
```diff
@@ -270,8 +270,10 @@ std::vector<double> ConvertFitFunctionForMuonTFAsymmetry::getNorms() {
 
   std::vector<double> norms(wsNames.size(), 0);
   auto colNames = table->getColumnNames();
-  auto wsNamesIndex = findName(colNames, "name");
-  auto normIndex = findName(colNames, "norm");
+  std::string colName = "name";
```
Comment (martyngigg) [inline Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp]: @AnthonyLim23 I think the problem here is that `findName` has the `name` variable taken by reference and not by value. If it is taken by value then you won't need these temporary variables (basically remove the `&` from `std::string &name`).

Comment (AnthonyLim23) [reply Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp]: but when I did that the cpp check was unstable

Comment (martyngigg) [reply Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp]: Ah I see. I guess in this case it actually makes little difference as you have to construct a string temporary. Maybe a simpler way, given the usage, is to change to `int findName(const std::vector<std::string>& colNames, const char* name)` and then we avoid the string allocation entirely.

**Pull request merged**